src/main/java/com/synopsys/integration/detect/tool/detector/inspectors/nuget/OnlineNugetInspectorLocator.java

```diff
@@ -8,51 +8,77 @@
 
-    public OnlineNugetInspectorLocator(final NugetInspectorInstaller nugetInspectorInstaller, final DirectoryManager directoryManager, @Nullable final String overrideVersion) {
+    public OnlineNugetInspectorLocator(NugetInspectorInstaller nugetInspectorInstaller, DirectoryManager directoryManager, @Nullable String overrideVersion,
+        InstalledToolManager installedToolManager, InstalledToolLocator installedToolLocator) {
         this.nugetInspectorInstaller = nugetInspectorInstaller;
         this.directoryManager = directoryManager;
         this.overrideVersion = overrideVersion;
+        this.installedToolManager = installedToolManager;
+        this.installedToolLocator = installedToolLocator;
     }
 
     @Override
     public File locateDotnet5Inspector() throws DetectableException {
-        return locateInspector(nugetInspectorInstaller::installDotNet5);
+        return locateInspector(nugetInspectorInstaller::installDotNet5, NUGET_INSPECTOR_5_INSTALLED_TOOL_JSON_KEY);
     }
 
     @Override
     public File locateDotnet3Inspector() throws DetectableException {
-        return locateInspector(nugetInspectorInstaller::installDotNet3);
+        return locateInspector(nugetInspectorInstaller::installDotNet3, NUGET_INSPECTOR_3_INSTALLED_TOOL_JSON_KEY);
     }
 
     @Override
     public File locateDotnetInspector() throws DetectableException {
-        return locateInspector(nugetInspectorInstaller::installDotNet);
+        return locateInspector(nugetInspectorInstaller::installDotNet, NUGET_INSPECTOR_INSTALLED_TOOL_JSON_KEY);
     }
 
     @Override
     public File locateExeInspector() throws DetectableException {
-        return locateInspector(nugetInspectorInstaller::installExeInspector);
+        return locateInspector(nugetInspectorInstaller::installExeInspector, NUGET_INSPECTOR_EXE_INSTALLED_TOOL_JSON_KEY);
     }
 
-    private File locateInspector(final ThrowingBiFunction<File, String, File, DetectableException> inspectorInstaller) throws DetectableException {
+    private File locateInspector(ThrowingBiFunction<File, String, File, DetectableException> inspectorInstaller, String inspectorKey) throws DetectableException {
+        File inspector;
+        Optional<File> cachedInstall = installedToolLocator.locateTool(inspectorKey);
         try {
-            final File nugetDirectory = directoryManager.getPermanentDirectory("nuget");
-            return inspectorInstaller.apply(nugetDirectory, overrideVersion);
-        } catch (final Exception e) {
+            File nugetDirectory = directoryManager.getPermanentDirectory("nuget");
+            inspector = inspectorInstaller.apply(nugetDirectory, overrideVersion);
+        } catch (Exception e) {
+            if (cachedInstall.isPresent()) {
+                return cachedInstall.get();
+            }
             throw new DetectableException("Unable to install the nuget inspector from Artifactory.", e);
         }
+        if (inspector == null) {
```

Comment:
The inspectorInstaller should fail with an exception rather than returning null. The return objects are not annotated as `Nullable` either.

I think this is a lot easier to follow
```
private File locateInspector(ThrowingBiFunction<File, String, File, DetectableException> inspectorInstaller, String inspectorKey) throws DetectableException {
    try {
        File nugetDirectory = directoryManager.getPermanentDirectory("nuget");
        File inspector = inspectorInstaller.apply(nugetDirectory, overrideVersion);
        installedToolManager.saveInstalledToolLocation(inspectorKey, inspector.getAbsolutePath());
        return inspector;
    } catch (Exception e) {
        return installedToolLocator.locateTool(inspectorKey).orElseThrow(() ->
            new DetectableException("Unable to install the nuget inspector from Artifactory.", e)
        );
    }
}
```